### PR TITLE
TokenCreate `handle()` implementation

### LINKED ISSUE #6355: Add `pureChecks` and `validateSemantics` implementations  for TokenCreateHandler

### LINKED ISSUE #6386: Implement `handle` for TokenCreateHandler
- [x] Add `CustomFeeValidator`
- [x] Add `TokenCustomFee` PBJ object
- [x] Implement `handle`
- [x] API for calling `mint`
- [x] Tests

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
Fixes #6386 
Fixes #6355 
- Added `TokenCreate` handle implementation
- Added validations for customFees to `CustomFeeValidator`
- To separate all `TokenCreate` validations, added `TokenCreateValidator` and `TokenAttributesValidator`

### CODE REVIEW

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java [hunk 1]
```diff
@@ -227,9 +236,9 @@ private void dispatchPrng(@NonNull final HandleContext handleContext) {
     private void finishUtilPrng(@NonNull final HandleContext handleContext) {
         final var recordBuilder = handleContext.recordBuilder(SingleTransactionRecordBuilder.class);
         final var entropy = recordBuilder.entropy();
-        if (entropy.kind() == EntropyOneOfType.PRNG_NUMBER) {
+        if (entropy.kind() == TransactionRecord.EntropyOneOfType.PRNG_NUMBER) {
             sideEffectsTracker.trackRandomNumber((Integer) entropy.value());
-        } else if (entropy.kind() == EntropyOneOfType.PRNG_BYTES) {
+        } else if (entropy.kind() == TransactionRecord.EntropyOneOfType.PRNG_BYTES) {
             sideEffectsTracker.trackRandomBytes(PbjConverter.asBytes((Bytes) entropy.value()));
```
Comment (rbair23) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java:242]: Why are these lines modified?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java [hunk 2]
```diff
@@ -76,14 +76,18 @@ public MonoTransactionDispatcher(
...
             case CONSENSUS_DELETE_TOPIC -> dispatchConsensusDeleteTopic(context);
             case CONSENSUS_SUBMIT_MESSAGE -> dispatchConsensusSubmitMessage(context);
+                /* ------------------ crypto -------------------------- */
             case CRYPTO_CREATE_ACCOUNT -> dispatchCryptoCreate(context);
             case CRYPTO_DELETE -> dispatchCryptoDelete(context);
             case CRYPTO_UPDATE_ACCOUNT -> dispatchCryptoUpdate(context);
-            case FREEZE -> dispatchFreeze(context);
+            case CRYPTO_APPROVE_ALLOWANCE -> dispatchCryptoApproveAllowance(context);
+            case CRYPTO_DELETE_ALLOWANCE -> dispatchCryptoDeleteAllowance(context);
+                /* ------------------ token -------------------------- */
```
Comment (rbair23) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java:90]: I love having these in there! Is this how spotless formats it? I wonder if we should use // instead of /* */

Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java:90]: I did not change them manually.  It could be one of these two - 
1. Because I touched that file spotless formatted it 
2. I changed IntelliJ setting to not use * imports

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java (hunk 1 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java:242]: I did not change them manually.  It could be one of these two - 
1. Because I touched that file spotless formatted it 
2. I changed IntelliJ setting to not use * imports

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java (hunk 2 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java:90]: I have added those lines and seems spotless moved them to center 😃 . Will change to use //

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+    /**
+     * All validations in handle needed for {@link TokenCreateTransactionBody} are done here
+     * @param context context
+     * @param accountStore account store
+     * @param op token create transaction body
+     * @param config tokens config
+     */
+    public void validate(
+            final HandleContext context,
```
Comment (mhess-swl) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java:103]: `@NonNull`s on these params

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            final ReadableAccountStore accountStore,
+            final TokenCreateTransactionBody op,
+            final TokensConfig config) {
+        final var treasury = accountStore.getAccountById(op.treasuryOrElse(AccountID.DEFAULT));
+        validateTrue(treasury != null, INVALID_TREASURY_ACCOUNT_FOR_TOKEN);
```
Comment (mhess-swl) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java:108]: Again, should this use `TokenHandlerHelper.getIfUsable(...)`?

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    public void validateAssociation(
+            @NonNull final EntitiesConfig entitiesConfig,
+            @NonNull final TokensConfig tokensConfig,
+            @NonNull final Account account,
+            @NonNull final Token token,
+            @NonNull final WritableTokenRelationStore tokenRelStore) {
+        validateFalse(
+                entitiesConfig.limitTokenAssociations()
+                        && account.numberAssociations() + 1 > tokensConfig.maxPerAccount(),
+                TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED);
```
Comment (mhess-swl) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java:186]: Again, seems there should be a check for tokens per account _and_ max tokens overall

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java (hunk 1 above)
Comment (mhess-swl) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java:242]: Spotless has done this to me too in the past

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java [hunk 6]
```diff
@@ -249,4 +264,16 @@ private void dispatchTokenBurn(@NonNull final HandleContext handleContext) {
...
+    private void dispatchCryptoApproveAllowance(final HandleContext handleContext) {
+        final var handler = handlers.cryptoApproveAllowanceHandler();
+        handler.handle(handleContext);
+        finishConsensusCreateTopic(handleContext);
+    }
+
+    private void dispatchCryptoDeleteAllowance(final HandleContext handleContext) {
+        final var handler = handlers.cryptoDeleteAllowanceHandler();
+        handler.handle(handleContext);
+        finishConsensusCreateTopic(handleContext);
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java]: No, fixed and pushed it in my changes I was making already.

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java [hunk 7]
```diff
@@ -40,16 +55,24 @@
...
-public class TokenCreateHandler implements TransactionHandler {
+public class TokenCreateHandler extends BaseTokenHandler implements TransactionHandler {
+    private final CustomFeesValidator customFeesValidator;
+    private final TokenCreateValidator tokenCreateValidator;
+
     @Inject
-    public TokenCreateHandler() {
-        // Exists for injection
+    public TokenCreateHandler(
+            @NonNull CustomFeesValidator customFeesValidator, @NonNull TokenCreateValidator tokenCreateValidator) {
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:64]: Done. Thanks

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java [hunk 8]
```diff
@@ -16,22 +16,37 @@
 
 package com.hedera.node.app.service.token.impl.handlers;
 
-import static com.hedera.hapi.node.base.ResponseCodeEnum.INVALID_AUTORENEW_ACCOUNT;
-import static com.hedera.hapi.node.base.ResponseCodeEnum.INVALID_CUSTOM_FEE_COLLECTOR;
-import static com.hedera.hapi.node.base.ResponseCodeEnum.INVALID_TREASURY_ACCOUNT_FOR_TOKEN;
+import static com.hedera.hapi.node.base.ResponseCodeEnum.*;
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:19]: Removed. Thanks

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java (hunk 2 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java]: Spotless moves them to center 😃 . Will change to use //. Thanks

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java [hunk 9]
```diff
@@ -66,8 +89,176 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
...
+        final var txn = context.body();
+        final var op = txn.tokenCreationOrThrow();
+        // Create or get needed config and stores
+        final var tokensConfig = context.configuration().getConfigData(TokensConfig.class);
+        final var accountStore = context.writableStore(WritableAccountStore.class);
+        final var tokenStore = context.writableStore(WritableTokenStore.class);
+        final var tokenRelationStore = context.writableStore(WritableTokenRelationStore.class);
+
+        /* Validate if the current token can be created */
+        validateTrue(tokenStore.sizeOfState() + 1 <= tokensConfig.maxNumber(), TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED);
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:108]: I think `TOKEN_MAX_SUPPLY_REACHED` is used only when we mint more tokens and totalSupply > maxSupply.

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java [hunk 10]
```diff
@@ -66,8 +89,176 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
...
+     * @param accountStore account store
+     * @param tokenRelStore token relation store
+     * @param requireCollectorAutoAssociation set of custom fees whose token denomination is set to sentinel value
+     */
+    private void associateAccounts(
+            final HandleContext context,
+            Token newToken,
+            WritableAccountStore accountStore,
+            WritableTokenRelationStore tokenRelStore,
+            Set<CustomFee> requireCollectorAutoAssociation) {
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:154]: Done. Thanks

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java [hunk 11]
```diff
@@ -66,8 +89,176 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
...
+    }
+
+    /**
+     * Create a new token with the given parameters.
+     * @param newTokenNum new token number
+     * @param op token creation transaction body
+     * @param resolvedExpiryMeta resolved expiry meta
+     * @return newly created token
+     */
+    private Token buildToken(long newTokenNum, TokenCreateTransactionBody op, ExpiryMeta resolvedExpiryMeta) {
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:184]: Done. Thanks

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/records/TokenCreateRecordBuilder.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package com.hedera.node.app.service.token.impl.records;
+
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.hapi.node.base.TokenID;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * A {@code RecordBuilder} specialization for tracking the side effects of a {@code CryptoCreate}
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/records/TokenCreateRecordBuilder.java:24]: Ah yes 😄

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java [hunk 13]
```diff
@@ -66,8 +89,176 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
...
+        final var givenExpiryMeta = getExpiryMeta(context.consensusNow().getEpochSecond(), op);
+        final var resolvedExpiryMeta = context.expiryValidator().resolveCreationAttempt(false, givenExpiryMeta);
+
+        // validate auto-renew account exists
+        if (resolvedExpiryMeta.autoRenewNum() != 0) {
+            final var id = AccountID.newBuilder()
+                    .accountNum(resolvedExpiryMeta.autoRenewNum())
+                    .build();
+            final var autoRenewAccount = accountStore.getAccountById(id);
+            validateTrue(autoRenewAccount != null, INVALID_AUTORENEW_ACCOUNT);
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:259]: 💯 Thanks

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:108]: Hmm, `CreateLogic` calls this:

```java
    public void assertCreatableTokens(final int n) {
        final var candidateNum = updatedNumTokens() + n;
        ensure(candidateNum <= dynamicProperties.maxNumTokens());
    }
```
It uses `maxNumTokens()`, not `maxTokensPerAccount()`. Am I missing something?

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java (hunk 5 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java:186]: Max tokens created check is in start of handle and will throw `MAX_ENTITIES_IN_PRICE_REGIME_HAVE_BEEN_CREATED`

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java (hunk 3 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java:103]: Done

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/CustomFeesValidator.java [hunk 14]
```diff
@@ -177,4 +198,50 @@ private boolean isFungibleCommon(@NonNull final TokenType tokenType) {
...
+        if (royaltyFee.hasFallbackFee()) {
+            final var fallbackFee = royaltyFee.fallbackFeeOrThrow();
+            if (fallbackFee.hasDenominatingTokenId()) {
+                final var denominatingTokenId = fallbackFee.denominatingTokenIdOrThrow();
+                validateTrue(denominatingTokenId.tokenNum() != 0, CUSTOM_FEE_DENOMINATION_MUST_BE_FUNGIBLE_COMMON);
+                validateExplicitTokenDenomination(
+                        fee.feeCollectorAccountId(), denominatingTokenId, tokenRelationStore, tokenStore);
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/CustomFeesValidator.java:245]: t seemed that way in monoservice

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/CustomFeesValidator.java [hunk 15]
```diff
@@ -177,4 +198,50 @@ private boolean isFungibleCommon(@NonNull final TokenType tokenType) {
...
+                feesWithCollectorsToAutoAssociate.add(fee);
+            } else {
+                validateExplicitTokenDenomination(
+                        fee.feeCollectorAccountId(), fixedFee.denominatingTokenId(), tokenRelationStore, tokenStore);
+            }
+        }
+    }
+
+    private void validateRoyaltyFeeForCreation(
+            final TokenType tokenType,
```
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/CustomFeesValidator.java:230]: Fixed. Thanks

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java]: I was looking at outdated code 🤦 good to go now!

**Review by mhess-swl — APPROVED** [review]

**Review by rbair23 — APPROVED** [review]
